# Statistic using Python for Data Science - Part 1

<p><b>About :</b> Penting mempelajari Statistika karena merupakan salah 1 fundamental dasar belajar data science. Pada modul ini, kamu akan dibekali dengan banyak teori dasar mengenai statistika.</p>

<p><b>Course :</b> <a href=https://academy.dqlab.id/main/package/practice/288?pf=0>Statistic using Python for Data Science - Part 1</a></p>

<h2>Module Content:</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li><a href="#statistika-dan-package-statiska">Pengenalan Statistika dan Package Statistika di Python</a></li>
        <ul>
            <li><a href="#konsep-dasar-statistika">Konsep Dasar Statistika</a></li>
            <li><a href="#statistik-dan-parameter">Statistik dan Parameter</a></li>
            <li><a href="#data-kualitatif-dan-data-kuantitatif">Data Kualitatif dan Data Kuantitatif</a></li>
            <li><a href="#tipe-statistika">Tipe Statistika</a></li>
            <li><a href="#skala-pengukuran">Skala Pengukuran</a></li>
            <li><a href="#package-statistika-python">Package Statistika di Python</a></li>
        </ul>
        <li><a href="#pengenalan-numpy-dan-pandas">Pengenalan Numpy dan Pandas</a></li>
        <li><a href="#ukuran-pusat">Ukuran Pusat</a></li>
        <ul>
            <li><a href="#mean">Mean</a></li>
            <li><a href="#median">Median</a></li>
            <li><a href="#modus">Modus</a></li>
            <li><a href="#kuantil">Kuantil</a></li>
            <li><a href="#agregasi-data">Agregasi Data</a></li>
        </ul>
        <li><a href="#ukuran-sebaran">Ukuran Sebaran</a></li>
        <ul>
            <li><a href="#proporsi-kategori">Proporsi Kategori</a></li>
            <li><a href="#rentang">Rentang</a></li>
            <li><a href="#variansi">Variansi</a></li>
            <li><a href="#deviasi-baku">Deviasi Baku</a></li>
        </ul>
        <li><a href="#korelasi">Korelasi</a></li>
        <ul>
            <li><a href="#korelasi-pearson">Korelasi Pearson</a></li>
            <li><a href="#korelasi-spearman">Korelasi Spearman</a></li>
            <li><a href="#korelasi-kendall">Korelasi Kendall</a></li>
            <li><a href="#korelasi-kausalitas">Interpretasi Nilai Korelasi dan Kaitannya dengan Kausalitas</a></li>
        </ul>
    </ul>
</div>

<h2><a name="statistika-dan-package-statiska"></a>Chapter 1: Pengenalan Statistika dan Package Statistika di Python</h2>

Secara definisi, statistika adalah bidang studi yang mempelajari bagaimana mengumpulkan dan menganalisa data. Jika mengambil definisi yang lebih lengkap, maka statistika dapat didefinisikan sebagai ilmu yang mempelajari cara pengumpulan data, menganalisis data untuk mendapatkan kesimpulan informasi sampai dapat dijadikan dasar pembuatan keputusan atau kebijakan.

Di dalam mempelajari statistika, akan mempelajari bagaimana caranya <b>mengorganisir</b> dan <b>membuat kesimpulan</b> dari data. Kegiatan mengorganisir dan membuay kesimpulan dari data disebut sebagai <b>statistika deskriptif</b>. Ada banyak cara bagaimana kita menyimpulkan suatu data, misalnya dengan membuat <b>grafik</b> atau dengan <b>angka</b>, misalnya mengambil suatu <b>rata-rata</b> dari data.

Selanjutnya ketika harus mengambil keputusan dari data yang sudah diolah, kemudian akan menggunakan <b>statistika inferensial</b> sehingga dapat mengambil keputusan yang benar dari data yang juga sudah diolah dengan benar.

<h3><a name="konsep-dasar-statistika"></a>Konsep Dasar Statistika</h3>

<b>Observasi:</b> Adalah suatu unit yang diukur dengan data. Beberapa contoh diantaranya adalah:
<li>Siswa</li>
<li>Warga negara</li>
<li>Hewan</li>
<li>Kendaraan</li>

<b>Populasi:</b> Adalah koleksi dari keseluruhan observasi. Beberapa contoh diantaranya adalah:
<li>Semua siswa yang ada di sekolah</li>
<li>Semua warga negara Indonesia,</li>
<li>Semua hewan yang ada di hutan lindung,</li>
<li>Semua kendaraan di Jakarta</li>

<b>Sampel:</b> Adalah sub koleksi dari populasi. Beberapa contoh diantaranya adalah:
<li>5 siswa dari masing-masing kelas di suatu sekolah</li>
<li>100 warga negara Indonesia yang diambil dari beberapa wilayah</li>
<li>Beberapa species hewan tertentu di suatu hutan lindung</li>
<li>3 jenis kendaraan di Jakarta</li>

<h3><a name="statistik-dan-parameter"></a>Statistik dan Parameter</h3>

Setiap kita mengkaji suatu permasalahan, biasanya kita menggunakan beberapa individu dari grup-grup tertentu. Misalnya, ketika kita ingin mengetahui prestasi siswa di suatu sekolah untuk mata pelajaran matematika, kita bisa saja menghitung semua nilai siswa lalu ambil nilai rata-ratanya. Atau dalam konteks pemasaran, kita ingin tahu untuk setiap segmen pasar, berapa besar pendapata yang bisa kita peroleh dari tiap segmen.

Namun ada kalanya kita hanya butuh sebagian kecil dari grup dikarenakan beberapa keterbatasan seperti biaya pengambilan data yang terlalu mahal atau bisa karena membutuhkan waktu analisis yang singkat karena harus membuat keputusan saat itu juga. Jika kasus seperti ini biasanya kita hanya mengambil sebagian kecil dari grup. Misalnya kita hanya mengambil 10 siswa dari tiap kelas untuk menghitung prestasi siswa atau kita hanya mengambil 20% dari tiap segmen pasar kita. Hal ini yang memunculkan dua istilah untuk kedua kasus ini: <b>parameter</b> dan <b>statistik</b>.

Parameter adalah <b>penjelasan atas populasi</b> sedangkan statistik hanya <b>menjelaskan sampel dari populasi</b>. Untuk kasus terkait segmentasi pasar, mengukur revenue dari setiap orang di dalam segmen adalah parameter sedangkan jika kita hanya mengambil 20% dari tiap segmen hal ini dikatakan sebagai statistik.

<h3><a name="data-kualitatif-dan-data-kuantitatif"></a>Data Kualitatif dan Data Kuantitatif</h3>

Terdapat dua kategori data yang terdapat pada populasi atau sampel, yaitu data kualitatif dan data kuantitatif.

<b>Data kualitatif</b> adalah data yang diperoleh dari <b>mengkategorikan atau menjelaskan</b> suatu atribut dari populasi atau sampel. Biasanya disebut juga sebagai data kategorik. Beberapa contoh data kualitatif misalnya warna rambut, golongan darah, nama jalan, nama produk yang digunakan dan lain sebagainya.Biasanya data kualitatif selalu disebutkan dalam bentuk kata ataupun simbol.

<b>Data kuantitatif</b> adalah data yang diperoleh dari <b>ukuran atau hitungan</b> di suatu populasi atau sampel. Data kuantitatif selalu berbentuk angka. Data seperti gaji, berat badan, populasi di suatu negara, dan jumlah pelanggan yang dimiliki suatu e-commerce termasuk data kuantitatif. Data kuantitatif sendiri dapat dibagi menjadi dua yaitu data <b>diskrit</b> dan data <b>kontinu</b>.
<ul>
    <li>Data yang diperoleh dari hasil perhitungan adalah data diskrit: jumlah pelanggan, jumlah produk, dan jumlah telefon yang diterima oleh customer service per harinya adalah beberapa contoh data <b>diskrit</b>.</li>
    <li>Data yang diperoleh dari hasil perhitungan namun dapat memuat rasio, desimal, atau bilangan irasional adalah data kontinu: berat badan, tinggi badan, waktu, dan gaji adalah beberapa contoh data <b>kontinu</b>.</li>
</ul>

<h3><a name="tipe-statistika"></a>Tipe Statistika</h3>

<img src="Resources/tipe_statistika.png" width=70%>

<b>Statistika deskriptif</b> digunakan untuk melakukan eksplorasi pada data, biasanya menggunakan teknik visualisasi data sebagai alat bantu untuk memahami bagaimana bentuk distribusi dan hubungan antara satu titik data dengan titik data lainnya.

<b>Statistika inferensial</b> digunakan untuk melakukan pemgambilan keputusan atas suatu simpulan terkait dengan data yang sedang dianalisa. Statistika inferensi memungkinkan kita mengambil kesimpulan dari suatu populasi dengan menggunakan sampel yang diambil dari populasi tersebut.

<h3><a name="skala-pengukuran"></a>Skala Pengukuran</h3>

Ada beberapa skala pengukuran sebagaimana ada beberapa cara kita dapat mengelompokkan objek yang ingin kita ukur. Beberapa tingkatan tersebut adalah:

<h4>Data Kategorikal</h4>

<img src="Resources/data_kategorikal.png" width=70%>

<ul>
<li><b>Skala Nominal:</b> adalah skala yang digunakan untuk mengkategorikan suatu objek pengamatan dengan objek pengamatan lainnya. Sebagai contoh yang termasuk skala nominal seperti gender, kategori barang, ras, status pernikahan dan lain sebagainya</li>
<li><b>Skala Ordinal:</b> adalah skala yang digunakan untuk mengurutkan suatu objek pengamatan dimana suatu titik pengamatan memiliki nilai yang lebih rendah atau tinggi dibanding nilai lainnya. Sebagai contoh yang termasuk skala nominal adalah kelas, jabatan, tingkat pendidikan, dan sebagainya</li>
</ul>

<h4>Data Numerikal</h4>

<img src="Resources/data_numerikal.png" width=70%>

<ul>
    <li><b>Skala interval:</b> adalah skala yang digunakan untuk tidak hanya untuk mengklasifikasikan maupun memberikan tingkatan pada suatu titik pengamatan, namun kita dapat mengukur seberapa besar nilai antara suatu titik pengamatan dengan titik pengamatan lainnya. Beberapa contoh yang termasuk skala interval diantaranya adalah suhu tubuh dan jarak.</li>
    <li><b>Skala Rasio:</b> memiliki kemiripan dengan skala interval, perbedaannya terletak pada nilai 0 pada skala rasio. Berbeda dengan skala interval yang tidak memiliki nilai 0 yang tidak pasti.</li>
</ul>

<h3><a name="package-statistika-python"></a>Package Statistika di Python</h3>

Untuk melakukan perhitungan statistika, kita dapat menggunakan beberapa package atau library berikut yang tersedia di Python. Diantaranya adalah:

<ul>
    <li><code>numpy</code>: digunakan untuk melakukan analisa data numerik dan perhitungan berbasis vektor atau matriks</li>
    <li><code>pandas</code>: digunakan untuk melakukan pengolahan data tabular</li>
    <li><code>matplotlib</code>: digunakan untuk melakukan ploting atau penggambaran grafik, dapat digunakan sebagai alat bantu dalam analisa data</li>
    <li><code>statsmodels</code>: digunakan untuk melakukan uji hipotesa, eksplorasi data maupun pemodelan statistika</li>
    <li><code>scipy</code>: digunakan untuk melakukan uji statistika, juga dapat digunakan untuk melakukan pemodelan statistika</li>
</ul>

Sebenarnya masih banyak lagi library python untuk pengolahan data statistika, namun untuk materi ini cukup 4 atau 5 library di atas saja yang akan kita gunakan.

<h2><a name="pengenalan-numpy-dan-pandas"></a>Chapter 2: Pengenalan Numpy dan Pandas</h2>

Kedua library ini, <code>numpy</code> dan <code>pandas</code>, adalah library yang umum digunakan untuk melakukan pengolahan data, mulai dari membaca data dalam format tertentu (<code>csv</code>, <code>xlsx</code>, <code>xls</code>, <code>tsv</code>) atau dari sumber tertentu (<code>RDBMS</code>, <code>No-SQL</code>), melakukan inspeksi (mengecek data yang hilang, inkonsistensi pada data) dan pengolahan data (transformasi nilai, encoding, normalisasi) sampai visualisasi data untuk membuat laporan atau mempersiapkan data untuk membuat model.

<b>Numpy</b> adalah library yang biasanya digunakan untuk manipulasi array atau vektor. Perhitungan yang melibatkan operasi pada objek berbentuk matriks, vektor atau bahkan multidimensi vektor (misalnya data gambar dengan skema warna RGB) dapat kita lakukan dengan menggunakan numpy.

<b>Pandas</b> adalah library yang biasanya digunakan untuk analisa data atau biasa disebut sebagai data wrangling. Biasanya data yang diolah oleh pandas berbentuk tabular atau tabel layaknya spreadsheet di Excel. Pandas menggunakan numpy sebagai back-end sehingga beberapa fungsi atau method dari numpy dapat digunakan di objek pandas.

In [2]:
import numpy as np
import pandas as pd

raw_data = pd.read_csv("https://dqlab-dataset.s3-ap-southeast-1.amazonaws.com/dataset_statistic.csv", sep=';')
print("Dimensi data:", raw_data.shape)
raw_data.head()

Dimensi data: (20, 9)


ID Pelanggan   Nama  Jenis Kelamin  Pendapatan Produk   Harga  Jumlah   \
0             1   Arif              1      600000      A  100000        4   
1             2   Dian              2     1200000      D  250000        4   
2             3  Dinda              2      950000      D  250000        3   
3             4  Fajar              1      400000      A  100000        2   
4             5    Ika              2     1200000      D  250000        4   

     Total  Tingkat Kepuasan  
0   400000                 2  
1  1000000                 2  
2   750000                 3  
3   200000                 3  
4  1000000                 2

In [3]:
raw_data.columns

Index(['ID Pelanggan', 'Nama', 'Jenis Kelamin', 'Pendapatan', 'Produk',
       'Harga', 'Jumlah ', 'Total', 'Tingkat Kepuasan'],
      dtype='object')

In [4]:
# check missing value
raw_data.isna().sum()

ID Pelanggan        0
Nama                0
Jenis Kelamin       0
Pendapatan          0
Produk              0
Harga               0
Jumlah              0
Total               0
Tingkat Kepuasan    0
dtype: int64

In [5]:
# melihat deskripsi statistik
raw_data.describe()

ID Pelanggan  Jenis Kelamin    Pendapatan          Harga    Jumlah   \
count      20.00000      20.000000  2.000000e+01      20.000000  20.000000   
mean       10.50000       1.700000  1.160000e+06  197500.000000   3.650000   
std         5.91608       0.470162  1.282842e+06   69726.910912   1.182103   
min         1.00000       1.000000  4.000000e+05  100000.000000   1.000000   
25%         5.75000       1.000000  6.750000e+05  150000.000000   3.000000   
50%        10.50000       2.000000  8.750000e+05  200000.000000   4.000000   
75%        15.25000       2.000000  1.125000e+06  250000.000000   4.250000   
max        20.00000       2.000000  6.450000e+06  300000.000000   5.000000   

              Total  Tingkat Kepuasan  
count  2.000000e+01         20.000000  
mean   7.100000e+05          2.050000  
std    3.338768e+05          0.887041  
min    2.000000e+05          1.000000  
25%    4.750000e+05          1.000000  
50%    6.750000e+05          2.000000  
75%    9.250000e+05          3.000000  
max    1.500000e+06          3.000000

In [6]:
# menghitung jumlah dari semua kolom
raw_data.sum()

ID Pelanggan                                                      210
Nama                ArifDianDindaFajarIkaIlhamIndraKartikaLestariL...
Jenis Kelamin                                                      34
Pendapatan                                                   23200000
Produk                                           ADDADBBEEEABCDCBCDAC
Harga                                                         3950000
Jumlah                                                             73
Total                                                        14200000
Tingkat Kepuasan                                                   41
dtype: object

In [7]:
# menghitung jumlah dari semua kolom bertipe data numerik saja
raw_data.sum(numeric_only=True)

ID Pelanggan             210
Jenis Kelamin             34
Pendapatan          23200000
Harga                3950000
Jumlah                    73
Total               14200000
Tingkat Kepuasan          41
dtype: int64

In [8]:
# menghitung jumlah dari kolom 'Harga' dan 'Pendapatan'
raw_data[['Harga', 'Pendapatan']].sum()

Harga          3950000
Pendapatan    23200000
dtype: int64

In [10]:
# manipulasi dataframe - memilih kolom
raw_data[['Jenis Kelamin', 'Pendapatan']].head()

Jenis Kelamin  Pendapatan
0              1      600000
1              2     1200000
2              2      950000
3              1      400000
4              2     1200000

In [14]:
# manipulasi dataframe - memilih baris
# mengambil data dari baris ke-0 sampai baris ke-(10-1) atau baris ke-9
raw_data[:10]

ID Pelanggan     Nama  Jenis Kelamin  Pendapatan Produk   Harga  Jumlah   \
0             1     Arif              1      600000      A  100000        4   
1             2     Dian              2     1200000      D  250000        4   
2             3    Dinda              2      950000      D  250000        3   
3             4    Fajar              1      400000      A  100000        2   
4             5      Ika              2     1200000      D  250000        4   
5             6    Ilham              1      800000      B  150000        4   
6             7    Indra              1      950000      B  150000        5   
7             8  Kartika              2     1100000      E  300000        3   
8             9  Lestari              2      800000      E  300000        2   
9            10      Lia              2     1700000      E  300000        5   

     Total  Tingkat Kepuasan  
0   400000                 2  
1  1000000                 2  
2   750000                 3  
3   200000                 3  
4  1000000                 2  
5   600000                 3  
6   750000                 1  
7   900000                 3  
8   600000                 1  
9  1500000                 1

In [15]:
# mengambil data dari baris ke-3 sampai baris ke-(5-1) atau baris ke-4
raw_data[3:5]

ID Pelanggan   Nama  Jenis Kelamin  Pendapatan Produk   Harga  Jumlah   \
3             4  Fajar              1      400000      A  100000        2   
4             5    Ika              2     1200000      D  250000        4   

     Total  Tingkat Kepuasan  
3   200000                 3  
4  1000000                 2

In [16]:
# mengambil data pada baris ke-1, ke-3 dan ke-10
raw_data.loc[[1,3,10]]

ID Pelanggan   Nama  Jenis Kelamin  Pendapatan Produk   Harga  Jumlah   \
1              2   Dian              2     1200000      D  250000        4   
3              4  Fajar              1      400000      A  100000        2   
10            11  Maria              2      600000      A  100000        4   

      Total  Tingkat Kepuasan  
1   1000000                 2  
3    200000                 3  
10   400000                 3

In [12]:
# manipulasi dataframe - memilih kolom dan baris
# Mengambil kolom 'Jenis Kelamin' dan 'Pendapatan' dan ambil baris ke-1 sampai ke-9
raw_data[['Jenis Kelamin', 'Pendapatan']][1:10]

Jenis Kelamin  Pendapatan
1              2     1200000
2              2      950000
3              1      400000
4              2     1200000
5              1      800000
6              1      950000
7              2     1100000
8              2      800000
9              2     1700000

In [13]:
# Mengambil kolom 'Harga' dan 'Tingkat Kepuasan' dan ambil baris ke-1, ke-10 dan ke-15
raw_data[['Harga', 'Tingkat Kepuasan']].loc[[1,10,15]]

Harga  Tingkat Kepuasan
1   250000                 2
10  100000                 3
15  150000                 1

<h2><a name="ukuran-pusat"></a>Chapter 3: Ukuran Pusat</h2>

Ukuran pusat (<i>measures of central tendency</i>) adalah statistika deskriptif yang dapat digunakan untuk membantu kita mengidentifikasi kasus-kasus tipikal di dalam sebuah sampel atau populasi.

Terdapat beberapa jenis ukuran pusat yang dapat digunakan untuk menganalisa data, diantaranya: <b>Mean</b> atau rerata, <b>Median</b>, <b>Modus</b>.

<h3><a name="mean"></a>Mean</h3>

Rata-rata atau mean adalah salah satu ukuran pusat yang nilainya diperoleh dengan cara menjumlahkan semua nilai titik data yang ada lalu dibagi oleh jumlah data. Secara matematis hal ini dapat dirumuskan sebagai berikut: 

$$
\overline{x} = \frac{x_1 + x_2 + ... + x_n}{n}
$$

dimana:
<ul>
    <li>x adalah rerata</li>
    <li>x1, x2, xn adalah titik data</li>
    <li>n adalah jumlah data</li>
</ul>

Kita dapat menghitung nilai rata-rata menggunakan method <code>.mean()</code> pada numpy atau pandas

In [18]:
# mengambil hanya data untuk produk 'A'
produk_A = raw_data[raw_data['Produk'] == 'A']
 
# menghitung rerata pendapatan menggunakan method .mean pada objek pandas DataFrame
print (produk_A['Pendapatan'].mean())
 
# menghitung rerata pendapatan menggunakan method .mean pada objek pandas DataFrame dengan numpy
print (np.mean(produk_A['Pendapatan']))

550000.0
550000.0


<h3><a name="median"></a>Median</h3>

Median adalah salah satu ukuran pusat yang nilainya terletak di tengah titik data. Sebagai gambaran, jika kita memiliki titik data bernilai 1, 2, 3, 4, 4, 5, 6 maka median dari sekumpulan titik data tersebut adalah 4. Namun jika kita memiliki titik data bernilai 1, 2, 3, 3 maka media dari sekumpulan titik data tersebut adalah:

$$
Median = \frac {2+3}{2} = 2.5
$$

Kita dapat menemukan nilai median dengan menggunakan method <code>.median()</code> pada numpy maupun pandas

In [21]:
# Hitung median dari pendapatan menggunakan pandas
print (produk_A['Pendapatan'].median())
 
# Hitung median dari pendapatan menggunakan numpy
print (np.median(produk_A['Pendapatan']))

600000.0
600000.0


<h3><a name="modus"></a>Modus</h3>

Modus didefinisikan sebagai data yang memiliki frekuensi kemunculan terbanyak/terbesar. Sebagai contoh, jika terdapat titik data seperti berikut: 1, 1, 1, 1, 2, 3, 3, 4 maka modus dari data tersebut adalah 1 karena 1 muncul sebanyak 4 kali, lebih banyak dibanding titik data lainnya.

Kita dapat menemukan nilai median dengan menggunakan method <code>.value_counts()</code> pada pandas

In [25]:
# Melihat jumlah dari masing-masing produk
raw_data['Produk'].value_counts()

D    5
A    4
B    4
C    4
E    3
Name: Produk, dtype: int64

<h3><a name="kuantil"></a>Kuantil</h3>

Kuantil adalah nilai-nilai data yang membagi data yang telah diurutkan sebelumnya menjadi beberapa bagian yang sama besar ukurannya. Beberapa ukuran fraktil ini diantaranya adalah:

<ul>
    <li>Kuartil: Adalah kuantil yang membagi data menjadi empat bagian sama besar. Kuartil ke-2 dari adalah median dari data tersebut.</li>
   <li>Desil: Adalah kuantil yang membagi data menjadi 10 bagian sama besar.</li>
   <li>Persentil: Adalah kuantil yang membagi data menjadi 100 bagian sama besar.</li>
</ul>

Untuk mencari fraktil dari data, kita dapat menggunakan method <code>.quantile</code> dari pandas atau numpy

In [26]:
# mencari median atau 50% dari data menggunakan pandas
print(raw_data['Pendapatan'].quantile(q = 0.5))
 
# mencari median atau 50% dari data menggunakan pandas
print(np.quantile(raw_data['Pendapatan'], q = 0.5))

875000.0
875000.0


<h3><a name="agregasi-data"></a>Agregasi Data</h3>

Ada kalanya kita ingin menghitung sekaligus beberapa ukuran, misalnya menghitung nilai mean sekaligus menghitung nilai median. Kita dapat melakukan kedua hal tersebut dengan menggunakan method <code>.agg()</code> pada objek pandas DataFrame

In [27]:
# menghitung rerata dan median 'Pendapatan' dan 'Harga'
raw_data[['Pendapatan', 'Harga']].agg([np.mean, np.median])

Pendapatan     Harga
mean     1160000.0  197500.0
median    875000.0  200000.0

In [28]:
# menghitung rerata dan median Pendapatan dan Harga dari tiap produk
raw_data[['Pendapatan', 'Harga', 'Produk']].groupby('Produk').agg([np.mean, np.median])

Pendapatan            Harga        
             mean   median    mean  median
Produk                                    
A          550000   600000  100000  100000
B          875000   875000  150000  150000
C          850000   900000  200000  200000
D         2100000  1200000  250000  250000
E         1200000  1100000  300000  300000

<h2><a name="ukuran-sebaran"></a>Chapter 4: Ukuran Sebaran</h2>

Ukuran sebaran (<i>measure of dispersion</i>) adalah statistika deskriptif yang digunakan untuk membantu kita memahami sebaran titik data di dalam sebuah sampel ataupun populasi.

Terdapat beberapa ukuran sebaran yang biasanya digunakan tergantung pada jenis atau tipe datanya, yaitu:

Tipe Data Nominal dan Ordinal
<li>Proporsi Kategori (<i>Categorical Proportion</i>)</li>
<li>Persen Proporsi (<i>Percent Proportion</i>)</li>
<li>Rasio Variasi (<i>Variation Ratio</i>)</li>

Tipe Data Interval dan Rasio
<li>Rentang (<i>Range</i>)</li>
<li>Variansi (<i>Variance</i>)</li>
<li>Deviasi Baku (<i>Standard Deviation</i>)</li>

<h3><a name="proporsi-kategori"></a>Proporsi Kategori</h3>

Proporsi kategori adalah ukuran sebaran yang paling sederhana dari ukuran sebaran pada data nomisal dan ordinal. Secara matematis dapat dirumuskan sebagai:

$$
Proporsi = \frac {N_{kategori ke-1}}{N_{total}}
$$

In [30]:
# cari proporsi tiap Produk
raw_data['Produk'].value_counts()/raw_data.shape[0]

D    0.25
A    0.20
B    0.20
C    0.20
E    0.15
Name: Produk, dtype: float64

<h3><a name="rentang"></a>Rentang</h3>

Rentang (<i>range</i>) adalah jarak antara nilai maksimum dengan nilai minimum. Semakin besar jarang antara nilai maksimum dan minimum semakin besar pula sebaran datanya. Secara matematis dapat dituliskan sebagai berikut:

$$
Range = max(x) - min(x)
$$

In [31]:
# Cari nilai rentang dari kolom 'Pendapatan'
raw_data['Pendapatan'].max() - raw_data['Pendapatan'].min()

6050000

<h3><a name="variansi"></a>Variansi</h3>

Variansi adalah ukuran sebaran pusat yang diperoleh dengan cara menghitung jarak antara tiap titik data pada sampel atau populasi dengan nilai mean. Secara matematis variansi dirumuskan sebagai berikut:

<b>Variansi Populasi</b>
$$
s^2 = \frac {\sum_{i=1}^{N} (X_i - \overline {X})^2}{N} = \frac {(X_1 - \overline {X})^2 + ... + (X_N - \overline {X})^2}{N}
$$
dimana N adalah jumlah data dan X1, ... XN adalah titik data pada sampel atau populasi

<b>Variansi Sampel</b>
$$
s^2 = \frac {\sum_{i=1}^{N} (X_i - \overline {X})^2}{N-1} = \frac {(X_1 - \overline {X})^2 + ... + (X_N - \overline {X})^2}{N-1}
$$

Untuk menghitung variansi kita dapat menggunakan method <code>.var()</code> dari numpy maupun pandas.

In [32]:
# menghitung variansi umur menggunakan method .var() dari pandas
print (raw_data['Pendapatan'].var())    # default variansi sampel
 
# menghitung variansi umur menggunakan method .var() dari numpy
print (np.var(raw_data['Pendapatan']))  # default variansi populasi

1645684210526.3157
1563400000000.0


<b style="color:red">Note:</b> Perhatikan bahwa nilai variansi keduanya berbeda. Hal ini karena secara default <b>pandas</b> menggunakan <b>variansi sampel</b> sedangkan <b>numpy</b> menggunakan <b>variansi populasi</b>. Untuk memperoleh hasil yang sama kita dapat menggunakan parameter <code>ddof=0</code> pada method <code>.var()</code> untuk memperoleh variansi populasi.

In [33]:
# mengatur variansi populasi dengan method `.var()` dari pandas
print (raw_data['Pendapatan'].var(ddof=0))  # menghitung variansei populasi

1563400000000.0


<h3><a name="deviasi-baku"></a>Deviasi Baku</h3>

Deviasi baku (<i>standard deviation</i>) adalah ukuran sebaran pusat yang diperoleh dengan cara menarik akar kuadrat dari hasil perhitungan variansi. Hal ini dilakukan karena nilai variansi umumnya memiliki nilai yang lebih besar daripada nilai aslinya sebagai efek dari pengkuadratan dan ini menjadikan variansi sulit untuk diinterpretasikan. Secara matematis deviasi baku dapat dirumuskan sebagi berikut:

<b>Deviasi Baku Populasi</b>
$$
s^2 = \sqrt {\frac {\sum_{i=1}^{N} (X_i - \overline {X})^2}{N}} = \sqrt {\frac {(X_1 - \overline {X})^2 + ... + (X_N - \overline {X})^2}{N}}
$$

<b>Deviasi Baku Sampel</b>
$$
s^2 = \sqrt {\frac {\sum_{i=1}^{N} (X_i - \overline {X})^2}{N-1}} = \sqrt {\frac {(X_1 - \overline {X})^2 + ... + (X_N - \overline {X})^2}{N-1}}
$$

Kita dapat menghitung deviasi baku menggunakan method <code>.std</code> dari numpy maupun pandas.

In [34]:
# menghitung deviasi baku sampel pendapatan menggunakan method std() dari pandas
print (raw_data['Pendapatan'].std())
 
# menghitung deviasi baku sampel pendapatan menggunakan method std() dari numpy
print (np.std(raw_data['Pendapatan'], ddof = 1))

1282842.2391417876
1282842.2391417876


<h2><a name="korelasi"></a>Chapter 5: Korelasi</h2>

Korelasi adalah salah satu metode statistika yang dapat digunakan untuk mengukur seberapa besar hubungan antara satu variabel dengan variabel lainnya. Sebagai contoh, misalnya mencari hubungan antara tinggi badan dengan berat badan, mencari hubungan antara gender dengan penghasilan dan masih banyak aplikasi penggunaan korelasi.

Terdapat beberapa metode yang dapat digunakan untuk menghitung korelasi antara sepasang variabel tergantung tipe dari sepasang variabel tersebut. Diantaranya adalah: <b>Korelasi Pearson</b>, <b>Korelasi Spearman</b>, <b>Korelasi Kendall</b>.

<h3><a name="korelasi-pearson"></a>Korelasi Pearson</h3>

Korelasi Pearson atau sering juga disebut sebagai Pearson's product moment correlation adalah pengukuran korelasi parametrik yang menghasilkan koefisien korelasi. Koefisien korelasi ini dapat digunakan untuk mengukur kekuatan hubungan atau asosiasi linier antara dua variabel. Artinya jika hubungan kedua variabel tidak linier maka koefisien korelasi Pearson tidak dapat digunakan untuk mengukur kekuatan hubungan antara kedua variabel.

Selain itu nilai dari koefisien pearson dapat digunakan untuk mengukur arah dari hubungan tersebut: positif atau negatif. Hubungan antar variabel dikatakan positif jika nilai salah satu variabel naik maka nilai variabel lainnya juga naik. Sebaliknya, hubungan antar variabel dikatakan negatif jika nilai salah satu variabel naik maka nilai variabelnya turun. Gambar berikut dapat menjelaskan maksud dari kekuatan dan arah dari korelasi antar kedua variabel.

<img src="Resources/korelasi_pearson.png">
Sumber Gambar: <a herf=https://www.mathsisfun.com>https://www.mathsisfun.com</a>

Beberapa asumsi yang harus dipenuhi untuk menggunakan korelasi Pearson diantaranya adalah:
<ol>
    <li>Nilai berskala interval/rasio</li>
    <li>Terdapat hubungan linier antara kedua variabel, yaitu kita dapat menggambarkan hubungan kedua variabel sebagai garis lurus.</li>
    <li>Kedua variabel berdistribusi normal</li>
    <li>Homoskedastis, atau data berdistribusi seragam dalam garis regresi</li>
</ol>

Secara matematis, korelasi pearson dapat dirumuskan sebagai berikut:

$$
r_{xy} = \frac {n \sum x_i y_i - \sum x_i \sum y_i}{\sqrt {n \sum x^2_i - (\sum x_i)^2} \sqrt {n \sum y^2_i - (\sum y_i)^2} }
$$

<h3><a name="korelasi-spearman"></a>Korelasi Spearman</h3>

Korelasi Spearman atau sering juga disebut sebagai Spearman's rank correlation adalah pengukuran korelasi non-parametrik. Artinya kita mencoba mengukur hubungan antara kedua variabel tanpa menghiraukan asumsi seperti distribusi dari kedua variabel dan asumsi lainnya. Secara kriteria memiliki kemiripan dengan korelasi Pearson walau korelasi Spearman bisa digunakan untuk data bertipe ordinal. Perbedaannya hanya terletak pada pengubahan data dalam bentuk ranking sebelum menghitung nilai korelasinya.

Secara matematis, korelasi Spearman dapat dihitung menggunakan rumus berikut:

$$
\rho = 1 - \frac {6 \sum d^2_i}{n(n^2 - 1)}
$$

Dimana:
<ul>
    <li>ρ adalah nilai korelasi Spearman</li>
    <li>di adalah nilai beda antara kedua variabel</li>
    <li>n adalah jumlah data</li>
</ul>

<h3><a name="korelasi-kendall"></a>Korelasi Kendall</h3>

Korelasi Kendall atau sering juga disebut juga sebagai Kendall's tank correlation atau korelasi Tau (τ) adalah pengukuran korelasi non-parametrik.

Secara matematis, korelasi Kendall dapat dihitung menggunakan rumus berikut:

$$
\tau = \frac {n_c - n_d}{\frac {1}{2}n(n-1)}
$$

Dimana: 
<ul>
    <li>nc adalah jumlah pasangan konkordan</li>
    <li>nd adalah jumlah pasangan diskordan</li>
    <li>n adalah jumlah data</li>
</ul>

Untuk menghitung korelasi antara kedua variabel, kita dapat menggunakan method <code>.corr()</code> dari pandas.

In [35]:
# menghitung korelasi dari setiap pasang variabel pada raw_data
raw_data.corr()

ID Pelanggan  Jenis Kelamin  Pendapatan     Harga   Jumlah   \
ID Pelanggan          1.000000       0.151375    0.110958 -0.028707  0.011289   
Jenis Kelamin         0.151375       1.000000    0.192849  0.457555 -0.104168   
Pendapatan            0.110958       0.192849    1.000000  0.322443  0.399825   
Harga                -0.028707       0.457555    0.322443  1.000000 -0.138883   
Jumlah                0.011289      -0.104168    0.399825 -0.138883  1.000000   
Total                -0.039968       0.238051    0.592044  0.645455  0.636097   
Tingkat Kepuasan     -0.245717      -0.088339   -0.312663 -0.380798  0.017568   

                     Total  Tingkat Kepuasan  
ID Pelanggan     -0.039968         -0.245717  
Jenis Kelamin     0.238051         -0.088339  
Pendapatan        0.592044         -0.312663  
Harga             0.645455         -0.380798  
Jumlah            0.636097          0.017568  
Total             1.000000         -0.268345  
Tingkat Kepuasan -0.268345          1.000000

Perhatikan bahwa hanya variabel bertipe numerik saja yang dihitung nilai korelasinya. Selain itu secara default method <code>.corr()</code> memiliki parameter method bernilai <b>pearson</b> sehingga nilai korelasi di atas adalah korelasi Pearson.
    
Untuk menggantinya menjadi korelasi Kendall atau korelasi Spearman kita cukup menambahkan parameter <code>method='kendall'</code> atau <code>method='spearman'</code> pada method <code>.corr()</code>.

In [36]:
# mencari korelasi 'kendall' untuk tiap pasang variabel
raw_data.corr(method='kendall')

ID Pelanggan  Jenis Kelamin  Pendapatan     Harga   Jumlah   \
ID Pelanggan          1.000000       0.126650   -0.054998 -0.005753  0.024016   
Jenis Kelamin         0.126650       1.000000    0.190245  0.415339 -0.090299   
Pendapatan           -0.054998       0.190245    1.000000  0.523053  0.501925   
Harga                -0.005753       0.415339    0.523053  1.000000 -0.098450   
Jumlah                0.024016      -0.090299    0.501925 -0.098450  1.000000   
Total                -0.065998       0.190245    0.988506  0.535078  0.501925   
Tingkat Kepuasan     -0.183817      -0.085796   -0.165588 -0.325659 -0.028923   

                     Total  Tingkat Kepuasan  
ID Pelanggan     -0.065998         -0.183817  
Jenis Kelamin     0.190245         -0.085796  
Pendapatan        0.988506         -0.165588  
Harga             0.535078         -0.325659  
Jumlah            0.501925         -0.028923  
Total             1.000000         -0.165588  
Tingkat Kepuasan -0.165588          1.000000

In [37]:
# mencari korelasi 'spearman' untuk tiap pasang variabel
raw_data.corr(method='spearman')

ID Pelanggan  Jenis Kelamin  Pendapatan     Harga   Jumlah   \
ID Pelanggan          1.000000       0.151375   -0.063711 -0.039149  0.046356   
Jenis Kelamin         0.151375       1.000000    0.219508  0.463635 -0.098864   
Pendapatan           -0.063711       0.219508    1.000000  0.640000  0.607110   
Harga                -0.039149       0.463635    0.640000  1.000000 -0.130749   
Jumlah                0.046356      -0.098864    0.607110 -0.130749  1.000000   
Total                -0.069779       0.219508    0.998470  0.646194  0.607110   
Tingkat Kepuasan     -0.238890      -0.090784   -0.192463 -0.378933 -0.023874   

                     Total  Tingkat Kepuasan  
ID Pelanggan     -0.069779         -0.238890  
Jenis Kelamin     0.219508         -0.090784  
Pendapatan        0.998470         -0.192463  
Harga             0.646194         -0.378933  
Jumlah            0.607110         -0.023874  
Total             1.000000         -0.192463  
Tingkat Kepuasan -0.192463          1.000000

<h3><a name=korelasi-kausalitas></a>Interpretasi Nilai Korelasi dan Kaitannya dengan Kausalitas</h3>

Perhatikan nilai korelasi Pearson dari raw_data berikut:

In [38]:
raw_data.corr()

ID Pelanggan  Jenis Kelamin  Pendapatan     Harga   Jumlah   \
ID Pelanggan          1.000000       0.151375    0.110958 -0.028707  0.011289   
Jenis Kelamin         0.151375       1.000000    0.192849  0.457555 -0.104168   
Pendapatan            0.110958       0.192849    1.000000  0.322443  0.399825   
Harga                -0.028707       0.457555    0.322443  1.000000 -0.138883   
Jumlah                0.011289      -0.104168    0.399825 -0.138883  1.000000   
Total                -0.039968       0.238051    0.592044  0.645455  0.636097   
Tingkat Kepuasan     -0.245717      -0.088339   -0.312663 -0.380798  0.017568   

                     Total  Tingkat Kepuasan  
ID Pelanggan     -0.039968         -0.245717  
Jenis Kelamin     0.238051         -0.088339  
Pendapatan        0.592044         -0.312663  
Harga             0.645455         -0.380798  
Jumlah            0.636097          0.017568  
Total             1.000000         -0.268345  
Tingkat Kepuasan -0.268345          1.000000

Perhatikan bahwa beberapa pasang variabel memiliki nilai korelasi yang positif maupun negatif. Selain itu terdapat nilai yang sangat kecil hingga mendekati nol, namun ada juga yang cukup besar diatas 0.5, misalnya korelasi antara 'Jumlah' dan 'Pendapatan'.

Besar kecilnya suatu nilai korelasi dari sepasang variabel menandakan seberapa kuat <b>hubungan linier</b> antara kedua variabel tersebut. Sebagai acuan untuk mengukur seberapa kuat korelasi sepasang variabel, kita dapat menggunakan ukuran berikut:

<ul>
    <li> 0 < |r| < 0.49: hubungan lemah</li>
    <li> 0.50 < |r| < 0.79: hubungan sedang</li>
    <li> 0.80 < |r| < 1: hubungan kuat</li>
</ul>

Berdasarkan kriteria di atas, kita dapat menilai bahwa hubungan antara variabel 'Total' dan 'Jumlah' hubungannya sedang dan positif (r = 0.636097) sedangkan hubungan antara 'Pendapatan' dan 'Tingkat Kepuasan' hubungannya lemah dan negatif (r = -0.088339).

Namun, perlu diperhatikan bahwa walaupun kita dapat menilai kuat hubungan antara kedua variabel, namun kita tidak bisa menentukan arah dari hubugan tersebut. Sebagai contoh, walaupun antara variabel 'Total' dan 'Jumlah' memiliki hubungan yang sedang, <b>tidak berarti</b> kita bisa mengetahui apakah kenaikan nilai 'Total' berefek positif terhadap 'Jumlah' atau sebaliknya.